Documentation/specs/runtime-configuration-file.md

```diff
@@ -0,0 +1,267 @@
+
+## `libraries` Section
+
+This section contains a union of all the dependencies found in the various targets, and contains common metadata for them. Specifically, it contains the `type`, as well as a boolean indicating if the library can be serviced (`servicable`, only for `package`-typed libraries) and a SHA-512 hash of the package file (`sha512`, only for `package`-typed libraries.
+
+**Open Question**: We could probably exclude projects from this set in order to reduce duplication. The main reason this is a separate section is because that's how the lock file is formatted and we want to try an keep this format the same if possible.
+
+## How the file is used
+
+The file is read by two different components:
+
+* `corehost` uses it to determine what to place on the TPA and Native Library Search Path lists, as well as what runtime settings to apply (GC type, etc.). See [the `corehost` spec](corehost.md).
+* `Microsoft.Extensions.DependencyModel` uses it to allow a running managed application to query various data about it's dependencies. For example:
+  * To find all dependencies that depend on a particular package (used by ASP.NET MVC and other plugin-based systems to identify assemblies that should be searched for possible plugin implementations)
+  * To determine the reference assemblies used by the application when it was compiled in order to allow runtime compilation to use the same reference assemblies (used by ASP.NET Razor to compile views)
+  * To determine the compilation settings used by the application in order to allow runtime compilation to use the same settings (also used by ASP.NET Razor views).
+  
+## Opt-In Compilation Data
+
+Some of the sections in the config file contain data used for runtime compilation. This data is not provided in the file by default. Instead, a project.json setting `preserveCompilationContext` must be set to true in order to ensure this data is added. Without this setting, the `compilationOptions` will not be present in the file, and the `targets` section will contain only the runtime target. For example, if the `preserveCompilationContext` setting was not present in the `project.json` that generated the above example, the file would only contain the following content:
+
+```json
+{
+    "runtimeConfig": {
+		"gcServer": true,
+		"gcConcurrent": false
+    },
+    "targets": {
+        "DNXCore,Version=v5.0/osx.10.10-x64": {
+            "MyApp/1.0": {
+                "type": "project",
+                "dependencies": {
+                    "AspNet.Mvc": "1.0.0"
+                }
+            },
+            "System.Foo/1.0.0": {
+                "type": "package",
+                "runtime": {
+                    "lib/dnxcore50/System.Foo.dll": { }
+                }
+            },
+            "System.Banana/1.0.0": {
+                "type": "package",
+                "dependencies": {
+                    "System.Foo": "1.0.0"
+                },
```

Comment:
No, the full closure is already present as the list of items in the target, you don't have to walk anything. The `dependencies` section is so that managed code can reconstitute the graph rather than using the flat list.